### ISSUE TITLE
Set-Cookie domain attribute actually needs leading dot to work

### ISSUE THREAD

**Issue by djc** [mdn/content#32050]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

### What specific section or headline is this issue about?

`Domain` attribute

### What information was incorrect, unhelpful, or incomplete?

> Contrary to earlier specifications, leading dots in domain names (`.example.com`) are ignored.

Despite the specification in RFC 6265, in our testing, it appears that current browsers (Chrome, Firefox, Safari) actually need the leading dot to make cookies from subdomains work in parent domains.

### What did you expect to see?

Cookies should be accessible across domains sharing a common suffix even when the `Domain` attribute does not specify a leading dot (as in `Domain = example.com`).

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/http/headers/set-cookie`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/http/headers/set-cookie/index.md
* Last commit: https://github.com/mdn/content/commit/28bfd9bd16b52e88f4f7c672a2e7e02f98748974
* Document last modified: 2024-01-10T10:30:39.000Z

</details>

**Comment by juangir** [comment]
Hi,
I've seen also this issue with most modern Browsers. 
But it seems that browsers have a lot stuff not being compliant with RFC 6265. [Here Ivan Nikulin
covers a lot of problems](http://inikulin.github.io/cookie-compat/).
I think in the MDN documentation should not state (as a fact) that cookies behave in this way. Also can be seen [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) in which states "_if you set Domain=mozilla.org, cookies are available on mozilla.org and its subdomains like developer.mozilla.org._" which in most cases is not true, and should be set to domain=.mozilla.org to be available to subdomains.

**Comment by dpchamps** [comment]
@djc do you have a repro of what you're observing?

**Comment by bsmth** [comment]
Thanks for reporting. I asked about this, and it's confusing because the **Domain** column in devtools shows what's in the cookie database's `Host`. The leading dot indicates that the cookie was set with an explicit domain attribute (`Domain=`). For example:

```http
Set-Cookie: <cookie-name>=<cookie-value>; Domain=<something>
```

So `example.com` and `.example.com` are equivalent (the leading dot is effectively ignored, per spec) and this is mostly a UI issue about displaying what's in the cookie DB. 

There's some more details on this here: https://bugzilla.mozilla.org/show_bug.cgi?id=1943081

**Comment by djc** [comment]
@bsmth I don't think that explains my experience -- we actually saw a difference in whether the cookie was used by the user agents, depending on whether the leading dot was present or not.

**Comment by bsmth** [comment]
Thanks for replying. Do you have an example of the tests you were running to reproduce this?

**Comment by dveditz** [comment]
@djc  are you seeing this difference in a `set-cookie:` header or when using `document.cookie`? Or are you talking about Developer Tools? The documentation on MDN is about the specification of the Set-Cookie syntax and behavior (also used when setting document.cookie).

The Developer Tools view is more directly showing you our internal representation and needs to be edited with that in mind. Internally we have combined the host/domain string with the "is this a host cookie or a domain cookie?" state into a single database field by using a leading '.' to signify "this is a domain cookie".

Ivan Nikulin's chart referenced by @juangir above appears to be from early 2016 judging by the browser versions. The chart is also about the original RFC 6265 when modern browsers currently implement "[rfc6265bis](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-19)", although the handling of leading dots in domain= should be the same.

This is working as specified according to the tests the browser vendors have agreed on to ensure compatibility: https://wpt.fyi/results/cookies/domain    If you've found a site where this is broken please send us a link or testcase. We would love to dissect it and make sure that case is covered by the WPT we're all using.

**Comment by Josh-Cena** [comment]
From the replies here it seems that our content is technically accurate. If there's anything we should change, with a reproduction example, please let us know.
